GH-40348: [Python] Add Python wrapper for VariableShapeTensor

rok opened the pull request:
### Rationale for this change

Once C++ implementation of VariableShapeTensor is complete (#38007) we want a Python wrapper for `VariableShapeTensor`.

### What changes are included in this PR?

This adds a Python wrapper for `VariableShapeTensor`.

### Are these changes tested?

This adds appropriate Python tests.

### Are there any user-facing changes?

Yes, a new extension and array type are exposed.
* GitHub Issue: #40348

AlenkaF reviewed (commented):
Small nits =) Thank you for working on this, takes lots of work!

rok commented on python/pyarrow/array.pxi:
```diff
@@ -4320,6 +4320,134 @@ cdef class FixedShapeTensorArray(ExtensionArray):
+    def from_numpy_ndarray(obj):
+        """
+        Convert a list of numpy arrays ndarrays to a variable shape tensor extension array.
+        The length of the list will become the length of the variable shape tensor array.
```
Sorry, there was a redundant "arrays" in there that didn't help for clarity, removed. So the idea is that each element (ndarray) of the original list becomes a row in the newly created array. Do you still think this needs to be rephrased?

AlenkaF replied:
No, I think it is good 👍

rok commented on python/pyarrow/array.pxi:
```diff
@@ -4320,6 +4320,134 @@ cdef class FixedShapeTensorArray(ExtensionArray):
+        >>> import pyarrow as pa
+        >>> import numpy as np
+        >>> ndarray_list = [
+        ...         np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32),
+        ...         np.array([[7, 8]], dtype=np.float32),
```
Added a check (`>>> assert len(ndarray_list) == arr.length`) below for clarity. Thanks!

AlenkaF commented:
The failures are connected.

rok commented:
@AlenkaF could you do a quick pass on this? I'm a little unsure about the `*row_splits` APIs, here's [more about the idea](http://tensorflow.org/api_docs/python/tf/RaggedTensor#from_row_splits).

rok committed:
GH-38007: [Python] Add VariableShapeTensor Python bindings

Add PyArrow bindings for the VariableShapeTensor extension type,
including VariableShapeTensorType, VariableShapeTensorArray, and
VariableShapeTensorScalar with support for converting to/from
NumPy tensors.

rok commented:
> Moving the `*row_splits` API related code into a separate issue and PR would be my suggestion. Otherwise, from first look, they sound useful. Might also be good to have a chat on the issue if it would be in fact useful for somebody?

Agreed, I moved these into a [separate PR](https://github.com/apache/arrow/pull/49747) and rebased here. I think this is ready for a more final review.

AlenkaF commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -3014,6 +3027,24 @@ cdef extern from "arrow/extension_type.h" namespace "arrow":
 
         shared_ptr[CArray] storage()
 
+cdef extern from "arrow/extension/variable_shape_tensor.h" namespace "arrow::extension" nogil:
```
Could this be moved little lower, after fixed shape tensor ref?

rok replied:
Done!

AlenkaF commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -3034,7 +3065,7 @@ cdef extern from "arrow/extension/uuid.h" namespace "arrow::extension" nogil:
 cdef extern from "arrow/extension/fixed_shape_tensor.h" namespace "arrow::extension" nogil:
     cdef cppclass CFixedShapeTensorType \
-            " arrow::extension::FixedShapeTensorType"(CExtensionType):
+            " arrow::extension::FixedShapeTensorType"(CExtensionType) nogil:
```
Is there a reason `nogil` is added here and used in the variable version also?

rok replied:
It's redundant, removing.

AlenkaF commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -908,6 +908,14 @@ cdef extern from "arrow/api.h" namespace "arrow" nogil:
             const shared_ptr[CBuffer] null_bitmap,
         )
 
+        @staticmethod
```
I am not sure where this is needed?

rok replied:
It does not appear to be used anywhere. Removing.

AlenkaF commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -1184,6 +1192,11 @@ cdef extern from "arrow/api.h" namespace "arrow" nogil:
         void set_chunksize(int64_t chunksize)
 
     cdef cppclass CTensor" arrow::Tensor":
+        CTensor(const shared_ptr[CDataType]& type,
```
Here also, I am not sure if we actually need this?

rok replied:
Removed.

AlenkaF commented on python/pyarrow/array.pxi:
```diff
@@ -4955,6 +4945,321 @@ cdef class Bool8Array(ExtensionArray):
+            VariableShapeTensorType ext_type
+
+        if isinstance(obj, np.ndarray):
+            raise TypeError("obj must be a sequence of numpy arrays")
```
I think the first check falls into the second one and could be omitted?

rok replied:
It sure seems so. Removed the first check.

AlenkaF commented on python/pyarrow/array.pxi:
```diff
@@ -4955,6 +4945,321 @@ cdef class Bool8Array(ExtensionArray):
+        array_ndim = arrays[0].ndim
+        arrow_type = from_numpy_dtype(base_dtype)
+
+        if value_type is not None and value_type != arrow_type:
```
Do we need to accept `value_type` if the input is not empty as it has to be same as the base type anyways? Not sure what are other options, but feels a bit funny as we are giving this option but it then has to be specific.

rok replied:
I've removed the option to create an array from an empty list (`VariableShapeTensorArray.from_numpy_ndarray([])` no longer works) and that removes the need for `value_type` and `ndim`, so these checks are no longer needed. If there turns out to be a need for calls like `VariableShapeTensorArray.from_numpy_ndarray([])` we can reintroduce this later.

AlenkaF replied:
OK, fair enough 👍

AlenkaF reviewed (commented):
Thank you for adding this @rok! Huge amount of work was needed here (and in the C++ implementation).

I did a first check but still need to go through most of the Cython code and all the tests.
As for the C++ permutation change - maybe that also fits better into a separate issue&PR?

rok committed:
review feedback

rok commented:
Thanks for the quick review @AlenkaF ! I addressed your comments, please do another pass whenever you can.

Regarding the C++ permutation change - I'll think about it a bit more and perhaps really spin it out. Again, I'm not entirely sure what to do yet, but it doesn't change the remainder of the review.

rok committed:
Revert the C++ change

rok committed:
fix permutation inference

raulcd reviewed (commented):
The failures on CI seem related. @rok Is this ready for review?

rok commented:
@raulcd yeah, it's related. I wanted to reason through strides/permutation logic one more time. The rest of the PR is ready for review.
